### PR TITLE
feat(backends): add pi coding-agent backend with ATIF trajectory parity

### PR DESCRIPTION
**Pull request by anderskev**
## Summary

- Adds the **Pi backend** (`PiBackend`) — a third code-agent backend alongside Claude and Codex, targeting the `pi` CLI (z.ai coding plan, GLM-5.2)
- Achieves full ATIF v1.6 trajectory parity with existing backends, including structured output extraction, session resumption, and cost/metrics emission
- Introduces a shared `extract_json()` helper for robust JSON extraction from prose-wrapped / markdown-fenced model output (GLM wraps JSON in prose)

## Motivation

The z.ai coding plan provides access to GLM-5.2 via the `pi` CLI. This backend lets daydream run review-fix loops against GLM alongside Claude and Codex, with the same trajectory contract and testing standard.

### Added
- `daydream/backends/pi.py` — full `PiBackend` implementation (587 lines): subprocess seam via `pi --mode json`, unified event mapping, continuation/resume tokens, read-only mode, skill-dir resolution
- `daydream/json_utils.py` — shared `extract_json()` helper for robust JSON extraction from prose-wrapped / markdown-fenced model output (used by `agent.py` and `pi.py`)
- `tests/test_backend_pi.py` — 60 tests covering event parsing, arg construction, structured output, session resumption, cost emission, cancellation, and the `PI_BASE_URL` models.json override
- `tests/test_json_utils.py` — 12 tests for the JSON extraction edge cases (fenced, prose-wrapped, nested braces, scan-forward-after-unparseable-span)
- `tests/harness/pi_replay.py` — replay harness mirroring `codex_replay.py` for contract parity tests
- `tests/contract/_loaders.py` — backend loader abstraction for the step-parity contract tests
- `tests/fixtures/pi_jsonl/` — 5 JSONL fixtures (simple_text, multi_turn, tool_use, structured_output, error_turn) + README

### Changed
- `daydream/backends/__init__.py` — registered `pi` in `VALID_BACKENDS` / `create_backend()`
- `daydream/config.py` — added pi phase-default model (`glm-5.2`), `DEFAULT_PI_MODEL`
- `daydream/agent.py` — wired `extract_json` into the structured-output fallback path; adapted return tuple for budget-cap (`aborted_reason`)
- `daydream/phases.py` — `phase_parse_feedback` now degrades gracefully (returns `[]` + warning instead of raising on malformed JSON)
- `daydream/cli.py` — pi backend wiring
- `daydream/trajectory.py` — pi event-type registration
- `tests/contract/test_backend_step_parity.py` — added pi-vs-claude parity cases; neutralized `_compare_steps` labels
- `tests/test_config.py` — pi phase-default coverage tests
- `CLAUDE.md` — pi backend architecture docs

### Fixed
- **Session resumption**: always pass `--session-id <uuid>` instead of `--no-session` (ephemeral sessions can't be resumed, breaking `phase_test_and_heal` retries)
- **Stderr deadlock**: merged `stderr=STDOUT` instead of `stderr=PIPE` (undrained pipe could deadlock when pi writes >64KB to stderr)
- **Structured output**: robust JSON extraction handles GLM's prose-wrapped / markdown-fenced JSON
- **`PI_BASE_URL`**: write a throwaway `models.json` in a temp agent dir (never mutates user's `~/.pi/agent/models.json`)
- **Codex live smoke test**: gated behind `DAYDREAM_CODEX_LIVE=1` opt-in (mirrors `DAYDREAM_PI_LIVE=1` — binary presence alone produces false reds when codex is unauthenticated)

### Removed
- `test_default_constants_still_exported` — `isinstance(str)` smoke test with no behavioral value

## Test Plan

- [x] Tests pass locally (`uv run pytest`) — 1654 passed, 3 skipped (live smoke opt-in)
- [x] Linting passes (`uv run ruff check`)
- [x] Type checking passes (`uv run mypy daydream`)
- [x] Daydream review-fix loop ran 3 rounds; all 24 findings addressed
- [x] LLM-artifact cleanup: 4-category parallel review, 15 findings (0 High), safe fixes applied
- [x] Full ATIF trajectory parity verified via contract tests
- [x] Documentation updated (CLAUDE.md)

## Checklist

- [x] Tests pass locally (`uv run pytest`)
- [x] Linting passes (`uv run ruff check`)
- [x] Type checking passes (`uv run mypy daydream`)
- [x] Documentation updated (if applicable)

### CODE REVIEW

File: daydream/backends/pi.py [hunk 1]
```diff
@@ -0,0 +1,587 @@
...
+            # Prefer the id pi reports in the session header (authoritative);
+            # fall back to the persistent id we passed via --session-id.
+            token_session = session_id or effective_session_id
+            yield ResultEvent(
+                structured_output=structured_result,
+                continuation=ContinuationToken(
+                    backend="pi",
+                    data={"session_id": token_session},
+                ),
+            )
```
[the remark this answers is not in the document]
Comment (anderskev) [reply daydream/backends/pi.py:488]: Fixed in 36f52bf. Added a  check after  that raises  before emitting terminal events. This prevents the false-green path where pi crashes without a  error event.

File: daydream/json_utils.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+                if ch == start_char:
+                    depth += 1
+                elif ch == end_char:
+                    depth -= 1
+                    if depth == 0:
+                        end_idx = i
+                        break
+            if end_idx == -1:
+                break  # unbalanced; try the next brace-type candidate
+            candidate = cleaned[start_idx : end_idx + 1]
```
[the remark this answers is not in the document]
Comment (anderskev) [reply daydream/json_utils.py:97]: Fixed in 36f52bf. Changed  to  so an unbalanced early span doesn't abandon the brace type entirely — we now scan forward for a later valid span of the same type.

File: tests/harness/pi_replay.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        async def readline(self) -> bytes:
+            try:
+                line = next(self._lines)
+                return (line + "\n").encode()
+            except StopIteration:
+                return b""
+
+    process = MagicMock()
+    process.stdout = _MockStdout()
+    process.stdin = MagicMock()
```
[the remark this answers is not in the document]
Comment (anderskev) [reply tests/harness/pi_replay.py]: Fixed in 36f52bf. Set  to match the real  semantics, so accidental stdin writes would surface immediately rather than being silently absorbed by a mock.
